### PR TITLE
‎Рамс Сергей. Задача 1. Вариант 23. Подсчет частоты символа в строке

### PR DESCRIPTION
**Pull request by 22380240**
Описание последовательной задачи:
Входные данные: строка и символ.
Подсчет числа вхождений символа в строку с помощью std::count.
Выходные данные: число вхождений символа в строку.

Описание параллельной задачи:
Входная строка равномерно делится между процессами. В случае если целочисленное деление длины строки на количество процессов невозможно, оставшиеся символы добавляются к подстроке последнего процесса. Каждый процесс считает число символов в своей подстроке. Полученные в каждом из процессов значения суммируются при помощи reduce.

### CODE REVIEW

File: tasks/mpi/rams_s_char_frequency/src/ops_mpi.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    delta = taskData->inputs_count[0] / world.size();
+    overflow = taskData->inputs_count[0] % world.size();
+  }
+
+  if (world.rank() == 0) {
+    input_ = std::string(reinterpret_cast<char*>(taskData->inputs[0]), taskData->inputs_count[0]);
+    target_ = *reinterpret_cast<char*>(taskData->inputs[1]);
+  }
+  broadcast(world, delta, 0);
+  broadcast(world, overflow, 0);
```
Comment (kmichaelk) [inline tasks/mpi/rams_s_char_frequency/src/ops_mpi.cpp:57]: Perhaps a cheaper way would be to send only the total number and calculate the side ones independently in each process

**Review by kmichaelk — APPROVED** [review]

File: tasks/mpi/rams_s_char_frequency/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
+// Copyright 2023 Nesterov Alexander
+#include <gtest/gtest.h>
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <vector>
+
+#include "mpi/rams_s_char_frequency/include/ops_mpi.hpp"
+
```
Comment (MoiseevArt) [inline tasks/mpi/rams_s_char_frequency/func_tests/main.cpp:9]: You can add some random tests

**Review by MoiseevArt — APPROVED** [review]

**Review by kmichaelk — APPROVED** [review]

File: tasks/mpi/rams_s_char_frequency/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    // Create Task
+    rams_s_char_frequency_mpi::TestMPITaskSequential testMpiTaskSequential(taskDataSeq);
+    ASSERT_EQ(testMpiTaskSequential.validation(), true);
+    testMpiTaskSequential.pre_processing();
+    testMpiTaskSequential.run();
+    testMpiTaskSequential.post_processing();
+
+    ASSERT_EQ(reference_out[0], global_out[0]);
+  }
+}
```
Comment (aobolensk) [inline tasks/mpi/rams_s_char_frequency/func_tests/main.cpp:264]: Please, add test for empty string

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/rams_s_char_frequency/func_tests/main.cpp (hunk 3 above)
Comment (22380240) [reply tasks/mpi/rams_s_char_frequency/func_tests/main.cpp:264]: Он уже есть - https://github.com/learning-process/ppc-2024-autumn/pull/65/files/53336be30925715fd310f1fd55bf8e58fb85bd1d#diff-3b660f176379fb8ec04535f8ffeee2d2ab7d72dd47d18ba33c1250badc24f6e9R109-R112

File: tasks/mpi/rams_s_char_frequency/src/ops_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  return true;
+}
+
+bool rams_s_char_frequency_mpi::TestMPITaskSequential::post_processing() {
+  internal_order_test();
+  reinterpret_cast<int*>(taskData->outputs[0])[0] = res;
+  return true;
+}
+
+bool rams_s_char_frequency_mpi::TestMPITaskParallel::pre_processing() {
```
Comment (allnes) [inline tasks/mpi/rams_s_char_frequency/src/ops_mpi.cpp:43]: please use all mpi functions in run()

**Review by allnes — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]
